src/m3ninx/index/segment/builder/sharded_fields_map.go

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package builder
+
+type shardedFieldsMap struct {
+	data []*fieldsMap
+}
+
+func newShardedFieldsMap(
+	numShards int,
+	shardInitialCapacity int,
+) *shardedFieldsMap {
+	data := make([]*fieldsMap, 0, numShards)
+	for i := 0; i < numShards; i++ {
+		data = append(data, newFieldsMap(fieldsMapOptions{
+			InitialSize: shardInitialCapacity,
+		}))
+	}
+	return &shardedFieldsMap{
+		data: data,
+	}
+}
+
+func (s *shardedFieldsMap) Get(k []byte) (*terms, bool) {
+	for _, fieldMap := range s.data {
+		if t, found := fieldMap.Get(k); found {
```

Comment:
Hm, can we perhaps just hash the `k` to work out which map to find?  

Otherwise we have to do 32 map accesses (on 32 core machine) vs checking just 1 map. This is done for every single field during a compaction to.

Reply:
Or just make the `Get(k)` take a shard too, i.e. `Get(shard int, k []byte)`

Reply:
Ah, I already have a `ShardedGet` method. The `(b *builder) Terms(field []byte)` call doesn't already have `shard` pre-computed so I added this regular `Get`. But you are right it looks like it's cheaper to compute the hash vs multiple map accesses. 

Here are some benchmark results:
```
notbdu @ Bos-MacBook-Pro.local m3 (bdu/concurrent-index) $ go test -v -bench . ./bench_test.go
goos: darwin
goarch: amd64
BenchmarkMapAccess-8    20000000                83.9 ns/op
BenchmarkHashing-8      200000000                6.21 ns/op
```